app/src/main/java/com/sschuraytz/charcoaldrawing/MainActivity.java

```diff
@@ -131,4 +158,121 @@ public void onUndoAvailable(boolean isAvailable) {
     public void onRedoAvailable(boolean isAvailable) {
         updateVisibility(isAvailable, redoButton);
     }
+
+    @RequiresApi(api = Build.VERSION_CODES.M)
+    public void checkVoicePermissions () {
+        if (ContextCompat.checkSelfPermission(
+                this, Manifest.permission.RECORD_AUDIO) == PackageManager.PERMISSION_GRANTED) {
+            //perform();
+        } else {
+           requestPermissions(
+               new String[] { Manifest.permission.RECORD_AUDIO},
+                1
+           );
+        }
+    }
+
+    public void setUpSpeechRecognizer() {
+        speechRecognizer = SpeechRecognizer.createSpeechRecognizer(this);
+        speechRecognizer.setRecognitionListener(recognitionListener);
+    }
+
+    public void setUpRecognitionListener() {
+        recognitionListener = new RecognitionListener() {
+            @Override
+            public void onReadyForSpeech(Bundle params) {
```

Comment:
You should move all this to its own class.

Reply:
I'm not sure if it's what you had in mind, but I created a Voice Listener similar to the UndoRedo listener to use in my VoiceCommand class.